Merge all patients 1

fredkingham opened the pull request:
Adds the managment commands `merge_all_patients` and `merge_patients_since` to merge all patients and just the patients in the last couple of days respectively.

This is done by the new `update_demographics.check_and_handle_upstream_merges_for_mrns` which takes in a bunch of MRNs and checks which and what need merging and then does the necessary.

Commit:
Changes the update demographics get_active_mrn_and_merged_mrn_data to create a cache.

Commit:
Adds an inital pass of merge_all_patients command

Commit:
Allow for merge comments being null or empty in update_demogrpahics

Commit:
Passes the cache in to the merge patient functionality

Commit:
Allows for there not to be a row for an mrn when looking at merges upstream

Commit:
Bug fixes update demographics so we don't always do the upstream query anyway

Commit:
Bug fix to the merge all patients script

Commit:
Formalises the cache

Commit:
Adds check_and_handle_upstream_merges_for_mrns

Adds the function
update_demographics.check_and_handle_upstream_merges_for_mrns that takes
in a list of mrns, checks to see which have been merged and process the
merges.

Commit:
Updates merge_all_patients to use check_and_handle_upstream_merges_for_mrns

Adds a function in update_demographics to return all the of the MRNs
of all patients who have been merged. Change the managment commadn to take these
MRNs and merge them with update_demographics.check_and_handle_upstream_merges_for_mrns

Commit:
Adds the merge_all_patients_since mgmt command

Adds the merge_all_patients_since that looks at the CRS_Patient_Masterfile for all
patients that have been updated in the last two days and merges them.

Commit:
Converts merge_all_patients_since to use spaces not tabs

Commit:
Renames merges_all_patients_since to merges_patients_since

Commit:
Adds facts for the merge_patients_since cmd

Adds facts and unit tests for the merge patients since mgmt command

Commit:
change merge_patients_since not to reraise

Changes merge_patients_since to not reraise this error as this is in
keeping with out other mgmt commands.

Commit:
Fixes/improves comments in the update_demographics.py

Commit:
Moves the one of merge all patient db logic into the mgmt command

Commit:
Changes merge all patients so that it does not load the patient.

Changes merge_all_patients so that it does not reload the patients
as the reload takes too long and we are going to resync all upstream
feeds.

Commit:
Renames cache to mrn_to_upstream_merge_data in update_demographics

Renames that ambiguous "cache" variable name to something that
explains what it is.

Commit:
Renames the cache variable in the loader to mrn_to_upstream_merge_data

Renames the cache variable in the intrahospital_api loader to
mrn_to_upstream_merge_data as this is what it is

Commit:
Replaces cache in update demographics tests with mrn_to_upstream_merge_data

the variable name of cache was replaced with mrn_to_upstream_merge_data
in update_demographics as its clearer. Lets make the change to the tests
as well.

fredkingham commented:
Moved back into in progress as it requires https://github.com/openhealthcare/elcid-rfh/pull/1440 to be merged.

Commit:
Fixes bug in merge_all_patients

Merge all patients was not calling the internal
check_and_handle_upstream_merges_for_mrns function and was instead
calling the one in update demographics.

We do not want this as we do not want it to run all the loaders.

fredkingham commented:
Needs a cron job

Commit:
Merge all patients should not bulk create at the end

We are out of memoring currently. So lets incrementally save rather
than do everything at the end.

Commit:
Remove the mrn to patient cache in merge all patients

The MRN to patient cache is causing the test server to out of memory so
lets remove it for the time being.

Commit:
Stops a naming conflict in merge_all_patients

We are importing merge_patient so lets call our function something
different

Commit:
Changes merge all patients use transactions and batch create

We are creating around 16k merged mrns which should be fine to create do
in a bulk create.

Put a transaction around the whole lot as we don't want some merged
patients to not have merged MRNs.

fredkingham commented:
Adds in the merge_dup_episodes command. We have a number of patients who have duplicate episode categories, the merge_patient command expects there to be only one episode of each category, this is so that it can know where to copy episode subrecords.

Commit:
Adds the merge patients since cron job to sync_demogrpahics

fredkingham commented:
Added the merge patients since job to the sync_demographics cron file. I've started adding cron jobs to the existing files rather than using new cron files each time. Roughly these groupings are based around packages, e.g. admissions has fetch_bed_status in it. There is no elcid cron job so I've put it in sync demographics

Commit:
Changes the merge all patients cron job to run every 6 hours

Commit:
Updates the doc strings on unit tests in test_update_demographics

Updates the doc strings for CheckAndHandleUpstreamMergesForMRNsTestCase
to make it clear what we are testing

Commit:
Adds additional checks in update_demographics tests

Just makes sure we are not calling merge_patient by mistake

Commit:
Fixes a bug in merge_patient not reloading patients

We were not reloading a patient in the hourly sync if they were not
being reloaded by merge_patient.

davidmiller commented on intrahospital_api/management/commands/merge_all_patients.py:
```diff
@@ -0,0 +1,166 @@
+    merge_patient.updates_statuses(new_patient)
+
+
+@transaction.atomic
```
I either don't think nested transactions work the way they're written in this file, or don't understand the intention...

fredkingham replied:
I don't _think_ it matters but we don't need 2 transaction atomics.

davidmiller commented on intrahospital_api/management/commands/merge_all_patients.py:
```diff
@@ -0,0 +1,166 @@
+
+    Filters those not related to elCID.
+
+    If they are inactive, creates a Patient for the
```
Is this accurate to the implementation isn't it filtered on active in the query?

Commit:
Removes transaction.atomic from merge_all_patients

We don't need two transaction atomics so lets remove one.

Commit:
Fixes a comment in merge_all_patients

Fixes the comment as in the cmd that function only takes active MRNs.

fredkingham commented:
Some stats..
#### Looking at all merge timestamps in everything merged upstream.
Since 2015 there have been 39 times where there have been more than 100 merges 
All of these were less than 300 merges except on 2/3/2022 where there were 9081 merges

#### Looking at fact timestamps 
Its run 152 times on test. 
104 of these times took less than 1 mins.
9 times over ten mins
3 times at 25 mins
1 time at 34 mins 
1 time at 40 mins.

Not sure what was happening in those situations, there's no massive jump in merges when there's a jump in time taken

Commit:
Refactors check_and_handle_upstream_merges_for_mrns

* Makes the check to see if an MRN is in elCID explicit.
* Make the removal of duplicates in the return of active_mrn to merged
  mrn dicts more explicit.
* Removes the mrn to patient map and does individual lookups instead.

Commit:
Merge branch 'cerner-merge-branch' into merge-all-patients-1

Commit:
Fixes unit tests in update_demographics

Commit:
Adds another unit test to CheckAndHandleUpstreamMergesForMRNsTestCase

*Adds another unit test for a case that was not covered by unit tests.
No non test needed to be editted as the case is already handled but it
should be testsed.

*Improves the doc strings on the unit tests

Commit:
Simplifies units tests for update_demographics

Given all the unit tests were mocking merge_patient and most were
mocking load_patient, handle this mocking at the class level.

Commit:
Changes check_and_handle_upstream_merges_for_mrns to send emails

If we are creating more than 300 merged MRNs in one hit, send an email
as this is more than we would expect and will want to check.

fredkingham commented:
This is going to be changed to send an email if we get an abnormally large amount back from the server.
It will also be changed to use https://github.com/openhealthcare/elcid-rfh/pull/1470 when merged.
Finally change the message sent out to reflect its the large amount of MergedMRNs we are worried about.

fredkingham commented:
The hours when there aremerges ...

8 3211
9 9084
10 7499
11 7999
12 6802
13 3878
14 15561
15 6298
16 5154
17 449
18 167

ie we don't need to run it over night.

Commit:
Merge branch 'cerner-merge-branch' into merge-all-patients-1

Commit:
Changes update_demographics to use utils.send_email

Changes update_demographics to send emails with the recently added
utils.send_email utility function

Commit:
Moves the too many mergedMRNs email and tweaks it in update_demographics

* Moves the email letting us know we are creating too many merged MRNs
  out of the already too large check_and_handle_upstream_merges_for_mrns
  function into its own function
* Tweaks the text of the email to tell the reader what is expected of them.

Commit:
Changes merge_patients_since to email if processing a lot of MRNs

We can handle 200000 rows but its more than we would expect to process in a month.
This makes the command send us an email if we are over this threshold,
just so we can log in and make sure the server is handling the increased
load as expected.

Commit:
Moves test_merge_patients_since into intrahospital_api

It was in admissions which was a mistake. It should be in
intrahospital_api where the management command merge_patients_since
lives.

fredkingham commented:
~~The 200,000 threshold came from the fact that that is over more than we would expect to return in a month (last month we returned 169,376).~~ <-- was the count for all (i.e. including non merged) MRNs. 

The total number of merged ='Y' patients updated in the last month is 4582

I test how it would cope, changing the command to run for the last month and it finished in under a minute

davidmiller commented on intrahospital_api/management/commands/merge_patients_since.py:
```diff
@@ -0,0 +1,38 @@
+from plugins.monitoring.models import Fact
+from elcid import utils
+
+UNEXPECTED_VOLUMES_THRESHOLD = 200000
```
I was thinking more like 500?
Given your test, maybe start with 1000, and put it in an actual setting so we don't need to redeploy to change it ?

fredkingham replied:
Sounds good, over the last month there have been 3 days were last_updated > 300, 1 instance where its been above 400 so I think 1000 is better than 500

davidmiller commented on intrahospital_api/management/commands/merge_patients_since.py:
```diff
@@ -0,0 +1,38 @@
+            upstream_merged_mrns = update_demographics.get_all_merged_mrns_since(since)
+            if len(upstream_merged_mrns) > UNEXPECTED_VOLUMES_THRESHOLD:
+                utils.send_email(
+                    f"{len(upstream_merged_mrns)} being processed by merge_patients_since",
```
You're missing the word MRN because you've read it in the variable name?

davidmiller commented on intrahospital_api/update_demographics.py:
```diff
@@ -220,13 +237,161 @@ def get_mrn_and_date_from_merge_comment(merge_comment):
+    logger.info('Generating merged MRNs')
+    to_create = []
+    for active_mrn, merged_dicts in active_mrn_and_merged_dicts:
+        active_patient = Patient.objects.filter(
```
May be missing something, but I think this spelling has you looping more times and setting up more datastructures than the way I would have spelled it...

```python
active_mrns_seen = set()
for mrn in mrns:
    active, merged_data = get_active_mrn_and_merged_mrn_data(mrn, mrn_to_upstream_merge_data)
    if active in active_mrns_seen:
        # The list passed into this function potentially contains all nodes in a merge graph
        continue 
    if not mrn_in_elcid(active):
        if not any(mrn_in_elcid(i['mrn']) for i in merged_data):
            continue # no node in this graph is part of the elCID cohort
    active_patient = Patient.objects.filter(demographics__hospital_unmber=active).first()
```

fredkingham replied:
Sure, I thought multiple for loops made it more readable, because each did one thing. I'm happy with either.

davidmiller commented on intrahospital_api/update_demographics.py:
```diff
@@ -220,13 +237,161 @@ def get_mrn_and_date_from_merge_comment(merge_comment):
+
+        # If we have patients that are inactive we need to do a merge.
+        if len(unmerged_patients) > 0:
+            if not active_patient:
```
Do you need to do this here as well as L317?

davidmiller commented on intrahospital_api/update_demographics.py:
```diff
@@ -220,13 +237,161 @@ def get_mrn_and_date_from_merge_comment(merge_comment):
+                    run_async=False
+                )
+            for unmerged_patient in unmerged_patients:
+                if active_patient:
```
Not sure I know what this if statement is for?

Commit:
Changes UNEXPECTED_VOLUMES_THRESHOLD to 1000

The previous unexpected volumes threshold was way too high. This is more
practical

Commit:
Fixes a typo in the alert email subject in merge_patients_since

Commit:
Refactors check_and_handle_upstream_merges_for_mrns to remove duplicate loops

Commit:
Merge branch 'cerner-merge-branch' into merge-all-patients-1

Commit:
Removes a redundant if statement in update_demographics

There should always be an active patient as this is created a couple of
lines above so the if statement is redundant.

davidmiller committed:
Refactor check_and_handle_upstream_merges_for_mrns

Removes some code after previous refactors, variable scoping, introduces a new function to reduce the method length.

davidmiller committed:
Update update_demographics.py

davidmiller committed:
Fixup:indentation

davidmiller committed:
Fixup:Variable name

davidmiller committed:
Fixup:variable names

davidmiller committed:
Fixup:Variable name

fredkingham committed:
Merge pull request #1471 from openhealthcare/davidmiller-patch-7

Refactor check_and_handle_upstream_merges_for_mrns

fredkingham commented:
So there is an issue with this...

When we call merge_patient.merge_patient we have not created MergedMRNs.

merge_patient.merge_patient implicitly assumes these MergedMRN for the patient exists when it calls load_patient. 

Without MergedMRNs the upstream data will not load in the data for the MergedRNs.

Commit:
renams mrn_in_elcid to is_mrn_in_elcid

Make it clearer the function returns a boolean

Commit:
Changes return values of update_demographics.get_or_create_active_patient

Changes get_or_create_active_patient to return True if the patient has
been created, otherwise False

Commit:
Renames merge_patient to merge_elcid_data and stops it calling load_patient

Stops merge_elcid_data calling load_patient and renames it so that its
clear that it is clear that its scope is elcid only.

Commit:
Refactors check_and_handle_upstream_merges_for_mrns

Changes it so that there is a function, merge_mrn_from_cerner, that we call per patient.
